src/ansible_compat/prerun.py

```diff
@@ -0,0 +1,427 @@
+    # it appears that is_dir() reports true instead, so we rely on exits().
+    target = pathlib.Path(options.project_dir).absolute()
+    if not link_path.exists() or os.readlink(link_path) != str(target):
+        if link_path.exists():
+            link_path.unlink()
+        link_path.symlink_to(target, target_is_directory=True)
+    _logger.info(
+        "Using %s symlink to current repository in order to enable Ansible to find the role using its expected full name.",
+        link_path,
+    )
+
+
+def _prepare_ansible_paths() -> None:
+    """Configure Ansible environment variables."""
+    library_paths: List[str] = []
+    roles_path: List[str] = []
+
+    for path_list, path in (
+        (library_paths, "plugins/modules"),
+        (library_paths, f"{options.cache_dir}/modules"),
+        (collection_list, f"{options.cache_dir}/collections"),
+        (roles_path, "roles"),
+        (roles_path, f"{options.cache_dir}/roles"),
+    ):
+        if path not in path_list and os.path.exists(path):
+            path_list.append(path)
+
+    _update_env('ANSIBLE_LIBRARY', library_paths)
+    _update_env(ansible_collections_path(), collection_list)
+    _update_env('ANSIBLE_ROLES_PATH', roles_path, default=ANSIBLE_DEFAULT_ROLES_PATH)
+
+
+def _update_env(varname: str, value: List[str], default: str = "") -> None:
+    """Update colon based environment variable if needed. by appending."""
+    if value:
+        orig_value = os.environ.get(varname, default=default)
+        if orig_value:
+            # Prepend original or default variable content to custom content.
+            value = [*orig_value.split(':'), *value]
+        value_str = ":".join(value)
+        if value_str != os.environ.get(varname, ""):
+            os.environ[varname] = value_str
+            _logger.info("Added %s=%s", varname, value_str)
+
+
+def ansible_config_get(key: str, kind: Type[Any] = str) -> Union[str, List[str], None]:
+    """Return configuration item from ansible config."""
+    env = os.environ.copy()
+    # Avoid possible ANSI garbage
+    env["ANSIBLE_FORCE_COLOR"] = "0"
+    # Avoid our own override as this prevents returning system paths.
+    colpathvar = ansible_collections_path()
+    if colpathvar in env:
+        env.pop(colpathvar)
+
+    config = subprocess.check_output(
+        ["ansible-config", "dump"], universal_newlines=True, env=env
+    )
+
+    if kind == str:
+        result = re.search(rf"^{key}.* = (.*)$", config, re.MULTILINE)
+        if result:
+            return result.groups()[0]
+    elif kind == list:
+        result = re.search(rf"^{key}.* = (\[.*\])$", config, re.MULTILINE)
+        if result:
+            val = eval(result.groups()[0])  # pylint: disable=eval-used
+            if not isinstance(val, list):
+                raise RuntimeError(f"Unexpected data read for {key}: {val}")
+            return val
+    else:
+        raise RuntimeError("Unknown data type.")
```

Comment:
Sounds like a `LookupError` would be more appropriate here.

Reply:
I would prefer to keep it like this as it assures that errors are nicely displayed by the consumers. Other exceptions may produce undesired stacktraces. We can debate that later but is better to avoid breaking the current "api".

Reply:
This exception is too generic. It's almost the same as just using `raise Exception`. You just don't do this.
Besides, you're leaking the implementation details from other abstraction layers.
Normally, you need to communicate exceptions that make sense for this function and not some wrapper that may or may not be there.
Ideally, you have to raise precise narrow exceptions and convert them to other exceptions at outer levels when it makes sense. It's something the caller has enough context to decide upon, not a low-level lib deep in the call stack.

Reply:
If you want compat with the old API, this could be an optional layer in this repo that some project may temporarily opt-in but it's mustn't be the default because you're shadowing what's happening. Otherwise, you make a dependency on the CLI of other projects while the dependency direction should be different. We shouldn't have cyclic deps.